NetworkProcessProxy should handle the case of busy looping service workers in a process containing web pages

youennf opened the pull request:
#### 52f9bcc211e160f6e3a8fb43a8af941be1892af8
<pre>
NetworkProcessProxy should handle the case of busy looping service workers in a process containing web pages
<a href="https://rdar.apple.com/138626537">rdar://138626537</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282091">https://bugs.webkit.org/show_bug.cgi?id=282091</a>

Reviewed by Ben Nham.

Add some logging for when a service worker becomes unresponsive.
Update the method names and IPC message to make it clearer that this is a service worker that is unresponsive, not the process itself.
Instead of requesting termination directly, we simply disable remote workers.
If disabling remote workers fails, the process will exit by itself.
If the process is hanging, a follow-up patch might make sure that the process gets killed as part of <a href="https://bugs.webkit.org/show_bug.cgi?id=282319.">https://bugs.webkit.org/show_bug.cgi?id=282319.</a>

When sending the notification that the service worker process is unresponsive to UIProcess, we make sure to close the SWContext connection.
This in turns stop all service workers running in that process.

* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::isProcessTerminationDelayEnabled const):
* Source/WebCore/workers/service/server/SWServerWorker.cpp:
(WebCore::SWServerWorker::startTermination):
(WebCore::SWServerWorker::callTerminationCallbacks):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::terminateDueToUnresponsiveness):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
(WebKit::NetworkProcessProxy::terminateUnresponsiveServiceWorkerProcesses): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didExceedCPULimit):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, SuspendAndTerminateWorker)):

Canonical link: <a href="https://commits.webkit.org/286269@main">https://commits.webkit.org/286269@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65024606bc7dd77d99bf541b3e123e25d4c3d5ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75240 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28080 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79701 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2464 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59051 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17294 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49227 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64632 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22129 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24824 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67719 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81186 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2570 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1590 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67302 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64632 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66593 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-default-size (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16602 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10537 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8698 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2531 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5336 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2556 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2160,14 +2160,14 @@ void WebProcessProxy::didExceedCPULimit()
 #if PLATFORM(MAC) && USE(RUNNINGBOARD)
     // This background WebProcess is using too much CPU so we try to suspend it if possible.
-    if (runningBoardThrottlingEnabled() && !throttler().isSuspended()) {
+    if (runningBoardThrottlingEnabled() && !throttler().isSuspended() && !isRunningServiceWorkers()) {
```
I was just looking at this code snippet last week and was wondering how it would make sense if the process was hosting a worker...

bnham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -2086,6 +2086,64 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+        return ![webView _hasServiceWorkerForegroundActivityForTesting] && ![webView _hasServiceWorkerBackgroundActivityForTesting];
+    }));
+
+    waitUntilEvaluatesToTrue([&] { return [webView _webProcessState] == _WKWebProcessStateSuspended; }, 500);
```
I would have expected this to take a long time to occur on Mac (due to the 8 min last visible activity timeout). Maybe we just don't go through that code path in the integration tests since the web view never goes through the visible => non-visible state transition.

youennf replied:
According my testing on Sequoia, it gets to suspended quickly.
On Ventura, it appears to not suspend at all.

bnham replied:
This may or may not be expected depending on the version of Ventura you're talking about, since `USE_RUNNINGBOARD` is a prereq for suspension, and it's only enabled in macOS >= 13.3.

youennf replied:
I am not sure, this happens on EWS bots, which are supposed to be above 13.3

bnham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -2086,6 +2086,64 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+    EXPECT_EQ([webView _webProcessState], _WKWebProcessStateSuspended);
+
+    // Process with service worker and page is suspended, let's terminate the service worker.
+    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^() {
```
This actually unsuspends all processes as a side effect in order to clear the memory cache (see `computeWebProcessAccessTypeForDataRemoval` and `WebProcessProxy::deleteWebsiteData`). So I think to test how termination interacts with process suspension, you need to specifically only delete the service worker data type.

youennf replied:
I'll check this.

youennf replied:
I updated the patch as it is better this way. That said, the web process continues to terminate properly the service worker in the test.

bnham commented on Source/WebCore/workers/service/server/SWServerWorker.cpp:
```diff
@@ -165,6 +169,8 @@ void SWServerWorker::callTerminationCallbacks()
 
 void SWServerWorker::terminationTimerFired()
 {
+    RELEASE_LOG_ERROR(ServiceWorker, "Terminating service worker %" PRIu64 " due to unresponsiveness", identifier().toUInt64());
```
👍

bnham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -2086,6 +2086,65 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+    TestWebKitAPI::Util::runFor(1_s);
+
+    // Let's verify the WKWebView did not crash.
+    [webView evaluateJavaScript:@"log('OK')" completionHandler: nil];
```
WKWebView auto-reloads on crash by default, maybe we should do a comparison of `_webProcessIdentifier` before and after?

youennf replied:
Ah ok, sounds good.

bnham commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -592,12 +592,10 @@ void NetworkProcessProxy::terminateWebProcess(WebCore::ProcessIdentifier webProc
-    if (auto process = WebProcessProxy::processForIdentifier(processIdentifier)) {
+    if (auto process = WebProcessProxy::processForIdentifier(processIdentifier))
         process->disableRemoteWorkers(RemoteWorkerType::ServiceWorker);
-        process->requestTermination(ProcessTerminationReason::ExceededCPULimit);
```
I am trying to make sure I understand the implications of removing this when it runs for a process that happens to be currently suspended (although maybe that would be almost impossible if we do [PR 35932](https://github.com/WebKit/WebKit/pull/35932)):

1. On the UIProcess side we'll clear out the worker info so that WebProcessProxy doesn't think it has a service worker anymore.
2. The WebContent process P won't be woken up to handle the termination event (since e.g. `WebSWContextManagerConnection::Close` is a one-way message), instead it'll handle the termination events once it resumes and gets to handling that message. Also, this process will not get terminated for failing responsiveness checks, since background responsiveness checks are disabled for suspended processes.
3. If another client needs this service worker, then the SW will start up in a different process and process P from (2) will learn about it through some queued IPC message.

youennf replied:
1 and 2 are correct (although PR 35932 will likely change this).
As of 3, the SW may go to whatever process is deemed appropriate.
If process P wakes up and asks for the service worker, it will likely be selected again.

bnham reviewed (approved):
LGTM assuming that my understanding of the new termination behavior is correct (see the inline comment).

youennf committed:
NetworkProcessProxy should handle the case of busy looping service workers in a process containing web pages
rdar://138626537
https://bugs.webkit.org/show_bug.cgi?id=282091

Reviewed by Ben Nham.

Add some logging for when a service worker becomes unresponsive.
Update the method names and IPC message to make it clearer that this is a service worker that is unresponsive, not the process itself.
Instead of requesting termination directly, we simply disable remote workers.
If disabling remote workers fails, the process will exit by itself.
If the process is hanging, a follow-up patch might make sure that the process gets killed as part of https://bugs.webkit.org/show_bug.cgi?id=282319.

When sending the notification that the service worker process is unresponsive to UIProcess, we make sure to close the SWContext connection.
This in turns stop all service workers running in that process.

* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::isProcessTerminationDelayEnabled const):
* Source/WebCore/workers/service/server/SWServerWorker.cpp:
(WebCore::SWServerWorker::startTermination):
(WebCore::SWServerWorker::callTerminationCallbacks):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::terminateDueToUnresponsiveness):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
(WebKit::NetworkProcessProxy::terminateUnresponsiveServiceWorkerProcesses): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didExceedCPULimit):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, SuspendAndTerminateWorker)):

Canonical link: https://commits.webkit.org/286269@main
